crates/ruff_python_formatter/src/expression/expr_bin_op.rs

```diff
@@ -290,3 +272,490 @@ impl NeedsParentheses for ExprBinOp {
+/// on a single line.
+///
+/// The formatting is recursive (with a depth of `O(operators)` where `operators` are operators with different precedences).
+///
+/// Comments before or after the first operand must be formatted by the caller because they shouldn't be part of the group
+/// wrapping the whole binary chain. This is to avoid that `b * c` expands in the following example because of its trailing comemnt:
+///
+/// ```python
+///
+/// ( a
+///     + b * c # comment
+///     + d
+/// )
+/// ```
+///
+///
+impl Format<PyFormatContext<'_>> for FlatBinaryExpressionSlice<'_> {
+    fn fmt(&self, f: &mut Formatter<PyFormatContext>) -> FormatResult<()> {
+        // Single operand slice
+        if let [OperandOrOperator::Operand(operand)] = &self.0 {
+            return operand.expression().format().fmt(f);
+        }
+
+        let mut last_operator: Option<OperatorIndex> = None;
+
+        let lowest_precedence = self.lowest_precedence();
+
+        for (index, operator_part) in self.operators() {
+            if operator_part.precedence() == lowest_precedence {
+                let left = self.between_operators(last_operator, index);
+                let right = self.after_operator(index);
+
+                let is_pow = operator_part.symbol.is_pow()
+                    && is_simple_power_expression(
+                        left.last_operand().expression(),
+                        right.first_operand().expression(),
+                    );
+
+                if let Some(leading) = left.first_operand().leading_binary_comments() {
+                    leading_comments(leading).fmt(f)?;
+                }
+
+                in_parentheses_only_group(&left).fmt(f)?;
+
+                if let Some(trailing) = left.last_operand().trailing_binary_comments() {
+                    trailing_comments(trailing).fmt(f)?;
+                }
+
+                if is_pow {
+                    in_parentheses_only_soft_line_break().fmt(f)?;
+                } else {
+                    in_parentheses_only_soft_line_break_or_space().fmt(f)?;
+                }
+
+                operator_part.fmt(f)?;
+
+                // Format the operator on its own line if the right side has any leading comments.
+                // if comments.has_leading(right.as_ref()) || !operator_comments.is_empty() {
+                if right
+                    .first_operand()
+                    .has_leading_comments(f.context().comments())
+                    || operator_part.has_trailing_comments()
+                {
+                    hard_line_break().fmt(f)?;
+                } else if !is_pow {
+                    space().fmt(f)?;
+                }
+
+                last_operator = Some(index);
+            }
+        }
+
+        // Format the last right side
+        // SAFETY: Last operator is guaranteed to be initialized because the slice must contain at least one operator.
```

Comment:
Doesn't it need to hold at least one _operand_? That's the debug assertion in the constructor.

Reply:
I extended the comment.